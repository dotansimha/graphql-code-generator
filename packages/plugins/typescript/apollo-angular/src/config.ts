import { RawClientSideBasePluginConfig } from '@graphql-codegen/visitor-plugin-common';

/**
 * @description This plugin generates Apollo services (`Query`, `Mutation` and `Subscription`) with TypeScript typings.
 *
 * It will generate a strongly typed Angular service for every defined query, mutation or subscription. The generated Angular services are ready to inject and use within your Angular component.
 *
 * It extends the basic TypeScript plugins: `@graphql-codegen/typescript`, `@graphql-codegen/typescript-operations` - and thus shares a similar configuration.
 *
 * To shed some more light regards this template, it's recommended to go through the this article: http://apollographql.com/docs/angular/basics/services.html , and to read the Code Generation with Apollo Angular: https://the-guild.dev/blog/apollo-angular-12
 */
export interface ApolloAngularRawPluginConfig extends RawClientSideBasePluginConfig {
  /**
   * @description Version of `apollo-angular` package
   * @default 2
   *
   * @exampleMarkdown
   * ```yml
   * config:
   *   apolloAngularVersion: 1
   * ```
   */
  apolloAngularVersion?: number;
  /**
   * @description Allows to define `ngModule` as part of the plugin's config so it's globally available.
   *
   * @exampleMarkdown
   * ```yml
   * config:
   *   ngModule: ./path/to/module#MyModule
   * ```
   */
  ngModule?: string;
  /**
   * @description Defined the global value of `namedClient`.
   *
   * @exampleMarkdown
   * ```yml
   * config:
   *   namedClient: 'customName'
   * ```
   */
  namedClient?: string;
  /**
   * @description Defined the global value of `serviceName`.
   *
   * @exampleMarkdown
   * ```yml
   * config:
   *   serviceName: 'MySDK'
   * ```
   */
  serviceName?: string;
  /**
   * @description Defined the global value of `serviceProvidedInRoot`.
   *
   * @exampleMarkdown
   * ```yml
   * config:
   *   serviceProvidedInRoot: false
   * ```
   */
  serviceProvidedInRoot?: boolean;
  /**
   * @description Define the Injector of the SDK class.
   *
   * @exampleMarkdown
   * ```yml
   * config:
   *   serviceProvidedIn: ./path/to/module#MyModule
   * ```
   */
  serviceProvidedIn?: string;
  /**
   * @description Set to `true` in order to generate a SDK service class that uses all generated services.
   * @default false
   */
  sdkClass?: boolean;
  /**
   * @description Allows to define a custom suffix for query operations.
   * @default GQL
   *
   * @exampleMarkdown
   * ```yml
   * config:
   *   querySuffix: 'QueryService'
   * ```
   */
  querySuffix?: string;
  /**
   * @description Allows to define a custom suffix for mutation operations.
   * @default GQL
   *
   * @exampleMarkdown
   * ```yml
   * config:
   *   mutationSuffix: 'MutationService'
   * ```
   */
  mutationSuffix?: string;
  /**
   * @description Allows to define a custom suffix for Subscription operations.
   * @default GQL
   *
   * @exampleMarkdown
   * ```yml
   * config:
   *   subscriptionSuffix: 'SubscriptionService'
   * ```
   */
  subscriptionSuffix?: 'GQL' | string;
  /**
   * @description Allows to define a custom Apollo-Angular package to import types from.
   * @default 'apollo-angular'
   */
  apolloAngularPackage?: 'apollo-angular' | string;
  /**
   * @description Add additional dependency injections for generated services
   * @default []
   *
   * @exampleMarkdown
   * ```yml
   * config:
   *   additionalDI
   *      - 'testService: TestService'
   *      - 'testService1': TestService1'
   * ```
   */
  additionalDI?: string[];
}
