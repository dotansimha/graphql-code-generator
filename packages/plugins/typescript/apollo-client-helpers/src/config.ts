export type ApolloClientHelpersConfig = {
  /**
   * @name useTypeImports
   * @type boolean
   * @default false
   * @description Will use `import type {}` rather than `import {}` when importing only types. This gives
   * compatibility with TypeScript's "importsNotUsedAsValues": "error" option
   *
   * @example
   * ```yml
   * config:
   *   useTypeImports: true
   * ```
   */
  useTypeImports?: boolean;
  /**
   * @name requireKeyFields
   * @type boolean
   * @default false
   * @description Remove optional sign from all `keyFields` fields.
   *
   */
  requireKeyFields?: boolean;
  /**
   * @name requirePoliciesForAllTypes
   * @type boolean
   * @default false
   * @description Remove optional sign from all generated keys of the root TypePolicy.
   *
   */
  requirePoliciesForAllTypes?: boolean;
};
