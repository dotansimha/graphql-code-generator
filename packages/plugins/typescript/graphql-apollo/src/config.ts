import { RawClientSideBasePluginConfig } from '@graphql-codegen/visitor-plugin-common';

/**
 * @description This plugin generates [`graphql-request`](https://npmjs.com/package/graphql-request) ready-to-use SDK, which is fully-typed.
 */

export interface RawGraphQLApolloPluginConfig extends RawClientSideBasePluginConfig {}
