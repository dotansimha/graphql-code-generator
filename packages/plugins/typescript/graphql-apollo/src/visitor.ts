import {
  ClientSideBasePluginConfig,
  ClientSideBaseVisitor,
  LoadedFragment,
} from '@graphql-codegen/visitor-plugin-common';
import autoBind from 'auto-bind';
import { camelCase } from 'change-case-all';
import { GraphQLSchema, OperationDefinitionNode, print } from 'graphql';
import { RawGraphQLApolloPluginConfig } from './config';

export interface GraphQLRequestPluginConfig extends ClientSideBasePluginConfig {
  rawRequest: boolean;
}

export class GraphQLApolloVisitor extends ClientSideBaseVisitor<
  RawGraphQLApolloPluginConfig,
  GraphQLRequestPluginConfig
> {
  private _operationsToInclude: {
    node: OperationDefinitionNode;
    documentVariableName: string;
    operationType: string;
    operationResultType: string;
    operationVariablesTypes: string;
  }[] = [];

  constructor(schema: GraphQLSchema, fragments: LoadedFragment[], rawConfig: RawGraphQLApolloPluginConfig) {
    super(schema, fragments, rawConfig, {});

    autoBind(this);

    const typeImport = this.config.useTypeImports ? 'import type' : 'import';

    this._additionalImports.push(
      `${typeImport} { ApolloClient, QueryOptions, SubscriptionOptions, MutationOptions } from '@apollo/client';`
    );
  }

  public OperationDefinition(node: OperationDefinitionNode) {
    const operationName = node.name?.value;

    if (!operationName) {
      // eslint-disable-next-line no-console
      console.warn(
        `Anonymous GraphQL operation was ignored in "typescript-graphql-request", please make sure to name your operation: `,
        print(node)
      );

      return null;
    }

    return super.OperationDefinition(node);
  }

  protected buildOperation(
    node: OperationDefinitionNode,
    documentVariableName: string,
    operationType: string,
    operationResultType: string,
    operationVariablesTypes: string
  ): string {
    this._operationsToInclude.push({
      node,
      documentVariableName,
      operationType,
      operationResultType,
      operationVariablesTypes,
    });

    return null;
  }

  public get sdkContent(): string {
    const sdkOperations = this._operationsToInclude.map(x => {
      const documentVariableName = x.documentVariableName;
      const optionType = GraphQLApolloVisitor.getApolloOperationOptionType(x.operationType);
      const generics = [x.operationResultType, x.operationVariablesTypes];

      const operationName = x.node.name.value;
      const genericParameter = x.operationType !== 'Mutation' ? [...generics].reverse() : [...generics];
      // the reason we're reversing: https://github.com/apollographql/apollo-client/issues/8537
      return `${camelCase(operationName)}${x.operationType}(options: Partial<${optionType}<${genericParameter.join(
        ', '
      )}>>) {
          return client.${GraphQLApolloVisitor.getApolloOperation(x.operationType)}<${generics.join(
        ', '
      )}>({...options, ${GraphQLApolloVisitor.getDocumentFieldName(x.operationType)}: ${documentVariableName}})
      }`;
    });
    return `export const getSdk = (client: ApolloClient<any>) => ({
      ${sdkOperations.join(',\n')}
    });
    export type SdkType = ReturnType<typeof getSdk>
`;
  }
  private static getDocumentFieldName(operationType: string): string {
    switch (operationType) {
      case 'Subscription':
        return 'query';
      case 'Mutation':
        return 'mutation';
      case 'Query':
        return 'query';
      default:
        throw new Error('unknown operation type: ' + operationType);
    }
  }
  private static getApolloOperation(operationType: string): string {
    switch (operationType) {
      case 'Subscription':
        return 'subscribe';
      case 'Mutation':
        return 'mutate';
      case 'Query':
        return 'query';
      default:
        throw new Error('unknown operation type: ' + operationType);
    }
  }

  private static getApolloOperationOptionType(operationType: string): string {
    switch (operationType) {
      case 'Subscription':
        return 'SubscriptionOptions';
      case 'Mutation':
        return 'MutationOptions';
      case 'Query':
        return 'QueryOptions';
      default:
        throw new Error('unknown operation type: ' + operationType);
    }
  }
}
