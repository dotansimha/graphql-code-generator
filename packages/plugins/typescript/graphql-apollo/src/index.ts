import { Types, PluginValidateFn, PluginFunction, oldVisit } from '@graphql-codegen/plugin-helpers';
import { GraphQLSchema, concatAST, Kind, FragmentDefinitionNode } from 'graphql';
import { RawClientSideBasePluginConfig, LoadedFragment } from '@graphql-codegen/visitor-plugin-common';
import { GraphQLApolloVisitor } from './visitor';
import { extname } from 'path';
import { RawGraphQLApolloPluginConfig } from './config';

export const plugin: PluginFunction<RawGraphQLApolloPluginConfig> = (
  schema: GraphQLSchema,
  documents: Types.DocumentFile[],
  config: RawGraphQLApolloPluginConfig
) => {
  const allAst = concatAST(documents.map(v => v.document));
  const allFragments: LoadedFragment[] = [
    ...(allAst.definitions.filter(d => d.kind === Kind.FRAGMENT_DEFINITION) as FragmentDefinitionNode[]).map(
      fragmentDef => ({
        node: fragmentDef,
        name: fragmentDef.name.value,
        onType: fragmentDef.typeCondition.name.value,
        isExternal: false,
      })
    ),
    ...(config.externalFragments || []),
  ];
  const visitor = new GraphQLApolloVisitor(schema, allFragments, config);
  const visitorResult = oldVisit(allAst, { leave: visitor });

  return {
    prepend: visitor.getImports(),
    content: [
      visitor.fragments,
      ...visitorResult.definitions.filter(t => typeof t === 'string'),
      visitor.sdkContent,
    ].join('\n'),
  };
};

export const validate: PluginValidateFn<any> = async (
  schema: GraphQLSchema,
  documents: Types.DocumentFile[],
  config: RawClientSideBasePluginConfig,
  outputFile: string
) => {
  if (extname(outputFile) !== '.ts' && extname(outputFile) !== 'tsx') {
    throw new Error(`Plugin "typescript-graphql-apollo" requires extension to be ".ts" or ".tsx"!`);
  }
};

export { GraphQLApolloVisitor };
