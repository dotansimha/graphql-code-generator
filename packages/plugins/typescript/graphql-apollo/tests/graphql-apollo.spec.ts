import { validateTs } from '@graphql-codegen/testing';
import { plugin } from '../src/index';
import { parse, buildClientSchema, GraphQLSchema } from 'graphql';
import { Types, mergeOutputs } from '@graphql-codegen/plugin-helpers';
import { plugin as tsPlugin, TypeScriptPluginConfig } from '@graphql-codegen/typescript';
import { plugin as tsDocumentsPlugin, TypeScriptDocumentsPluginConfig } from '@graphql-codegen/typescript-operations';
import { RawGraphQLApolloPluginConfig } from '../src/config';

describe('apollo-client', () => {
  const schema = buildClientSchema(require('../../../../../dev-test/githunt/schema.json'));
  const basicDoc = parse(/* GraphQL */ `
    query feed {
      feed {
        id
        commentCount
        repository {
          owner {
            avatar_url
          }
        }
      }
    }

    subscription ListenToComments($name: String) {
      commentAdded(repoFullName: $name) {
        id
      }
    }

    mutation MutationOperation($comment: Comment!) {
      submitComment(Comment: $comment) {
        id
      }
    }

    query feed2($v: String!) {
      feed {
        id
      }
    }

    query feed3($v: String) {
      feed {
        id
      }
    }

    query feed4($v: String! = "TEST") {
      feed {
        id
      }
    }
  `);

  const validate = async (
    content: Types.PluginOutput,
    config: TypeScriptPluginConfig & TypeScriptDocumentsPluginConfig & RawGraphQLApolloPluginConfig,
    docs: Types.DocumentFile[],
    pluginSchema: GraphQLSchema,
    usage: string
  ) => {
    const m = mergeOutputs([
      await tsPlugin(pluginSchema, docs, config, { outputFile: '' }),
      await tsDocumentsPlugin(pluginSchema, docs, config),
      content,
      usage,
    ]);

    validateTs(m);

    return m;
  };

  describe('sdk', () => {
    it('Should support useTypeImports', async () => {
      const config = { useTypeImports: true };
      const docs = [{ location: '', document: basicDoc }];
      const result = (await plugin(schema, docs, config, {
        outputFile: 'graphql.ts',
      })) as Types.ComplexPluginOutput;

      const usage = `
async function test() {
  const Client = require('@apollo/client').ApolloClient;
  const client = new Client('');
  const sdk = getSdk(client);

  await sdk.feed();
  await sdk.feed3();
  await sdk.feed4();

  const result = await sdk.feed2({ v: "1" });

  if (result.feed) {
    if (result.feed[0]) {
      const id = result.feed[0].id
    }
  }
}`;
      const output = await validate(result, config, docs, schema, usage);

      expect(output).toMatchSnapshot();
    });

    it('Should generate Sdk correctly', async () => {
      const config = {};
      const docs = [{ location: '', document: basicDoc }];
      const result = (await plugin(schema, docs, config, {
        outputFile: 'graphql.ts',
      })) as Types.ComplexPluginOutput;

      const usage = `
async function test() {
  const Client = require('@apollo/client').ApolloClient;
  const client = new Client('');
  const sdk = getSdk(client);

  await sdk.feed();
  await sdk.feed3();
  await sdk.feed4();

  const result = await sdk.feed2({ v: "1" });

  if (result.feed) {
    if (result.feed[0]) {
      const id = result.feed[0].id
    }
  }
}`;
      const output = await validate(result, config, docs, schema, usage);

      expect(output).toMatchSnapshot();
    });
  });
});
