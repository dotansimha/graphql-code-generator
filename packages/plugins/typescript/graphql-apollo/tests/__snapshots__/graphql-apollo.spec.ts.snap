// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apollo-client sdk Should generate Sdk correctly 1`] = `
"export type Maybe<T> = T | null;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
import { ApolloClient, QueryOptions, SubscriptionOptions, MutationOptions } from '@apollo/client';
import gql from 'graphql-tag';
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
};

export type Query = {
  __typename?: 'Query';
  /** A feed of repository submissions */
  feed?: Maybe<Array<Maybe<Entry>>>;
  /** A single entry */
  entry?: Maybe<Entry>;
  /** Return the currently logged in user, or null if nobody is logged in */
  currentUser?: Maybe<User>;
};


export type QueryFeedArgs = {
  type: FeedType;
  offset?: Maybe<Scalars['Int']>;
  limit?: Maybe<Scalars['Int']>;
};


export type QueryEntryArgs = {
  repoFullName: Scalars['String'];
};

/** A list of options for the sort order of the feed */
export enum FeedType {
  /** Sort by a combination of freshness and score, using Reddit's algorithm */
  Hot = 'HOT',
  /** Newest entries first */
  New = 'NEW',
  /** Highest score entries first */
  Top = 'TOP'
}

/** Information about a GitHub repository submitted to GitHunt */
export type Entry = {
  __typename?: 'Entry';
  /** Information about the repository from GitHub */
  repository: Repository;
  /** The GitHub user who submitted this entry */
  postedBy: User;
  /** A timestamp of when the entry was submitted */
  createdAt: Scalars['Float'];
  /** The score of this repository, upvotes - downvotes */
  score: Scalars['Int'];
  /** The hot score of this repository */
  hotScore: Scalars['Float'];
  /** Comments posted about this repository */
  comments: Array<Maybe<Comment>>;
  /** The number of comments posted about this repository */
  commentCount: Scalars['Int'];
  /** The SQL ID of this entry */
  id: Scalars['Int'];
  /** XXX to be changed */
  vote: Vote;
};


/** Information about a GitHub repository submitted to GitHunt */
export type EntryCommentsArgs = {
  limit?: Maybe<Scalars['Int']>;
  offset?: Maybe<Scalars['Int']>;
};

/**
 * A repository object from the GitHub API. This uses the exact field names returned by the
 * GitHub API for simplicity, even though the convention for GraphQL is usually to camel case.
 */
export type Repository = {
  __typename?: 'Repository';
  /** Just the name of the repository, e.g. GitHunt-API */
  name: Scalars['String'];
  /** The full name of the repository with the username, e.g. apollostack/GitHunt-API */
  full_name: Scalars['String'];
  /** The description of the repository */
  description?: Maybe<Scalars['String']>;
  /** The link to the repository on GitHub */
  html_url: Scalars['String'];
  /** The number of people who have starred this repository on GitHub */
  stargazers_count: Scalars['Int'];
  /** The number of open issues on this repository on GitHub */
  open_issues_count?: Maybe<Scalars['Int']>;
  /** The owner of this repository on GitHub, e.g. apollostack */
  owner?: Maybe<User>;
};

/** A user object from the GitHub API. This uses the exact field names returned from the GitHub API. */
export type User = {
  __typename?: 'User';
  /** The name of the user, e.g. apollostack */
  login: Scalars['String'];
  /** The URL to a directly embeddable image for this user's avatar */
  avatar_url: Scalars['String'];
  /** The URL of this user's GitHub page */
  html_url: Scalars['String'];
};

/** A comment about an entry, submitted by a user */
export type Comment = {
  __typename?: 'Comment';
  /** The SQL ID of this entry */
  id: Scalars['Int'];
  /** The GitHub user who posted the comment */
  postedBy: User;
  /** A timestamp of when the comment was posted */
  createdAt: Scalars['Float'];
  /** The text of the comment */
  content: Scalars['String'];
  /** The repository which this comment is about */
  repoName: Scalars['String'];
};

/** XXX to be removed */
export type Vote = {
  __typename?: 'Vote';
  vote_value: Scalars['Int'];
};

export type Mutation = {
  __typename?: 'Mutation';
  /** Submit a new repository, returns the new submission */
  submitRepository?: Maybe<Entry>;
  /** Vote on a repository submission, returns the submission that was voted on */
  vote?: Maybe<Entry>;
  /** Comment on a repository, returns the new comment */
  submitComment?: Maybe<Comment>;
};


export type MutationSubmitRepositoryArgs = {
  repoFullName: Scalars['String'];
};


export type MutationVoteArgs = {
  repoFullName: Scalars['String'];
  type: VoteType;
};


export type MutationSubmitCommentArgs = {
  repoFullName: Scalars['String'];
  commentContent: Scalars['String'];
};

/** The type of vote to record, when submitting a vote */
export enum VoteType {
  Up = 'UP',
  Down = 'DOWN',
  Cancel = 'CANCEL'
}

export type Subscription = {
  __typename?: 'Subscription';
  /** Subscription fires on every comment added */
  commentAdded?: Maybe<Comment>;
};


export type SubscriptionCommentAddedArgs = {
  repoFullName: Scalars['String'];
};
export type FeedQueryVariables = Exact<{ [key: string]: never; }>;


export type FeedQuery = { __typename?: 'Query', feed?: Maybe<Array<Maybe<{ __typename?: 'Entry', id: number, commentCount: number, repository: { __typename?: 'Repository', owner?: Maybe<{ __typename?: 'User', avatar_url: string }> } }>>> };

export type ListenToCommentsSubscriptionVariables = Exact<{
  name?: Maybe<Scalars['String']>;
}>;


export type ListenToCommentsSubscription = { __typename?: 'Subscription', commentAdded?: Maybe<{ __typename?: 'Comment', id: number }> };

export type MutationOperationMutationVariables = Exact<{
  comment: Comment;
}>;


export type MutationOperationMutation = { __typename?: 'Mutation', submitComment?: Maybe<{ __typename?: 'Comment', id: number }> };

export type Feed2QueryVariables = Exact<{
  v: Scalars['String'];
}>;


export type Feed2Query = { __typename?: 'Query', feed?: Maybe<Array<Maybe<{ __typename?: 'Entry', id: number }>>> };

export type Feed3QueryVariables = Exact<{
  v?: Maybe<Scalars['String']>;
}>;


export type Feed3Query = { __typename?: 'Query', feed?: Maybe<Array<Maybe<{ __typename?: 'Entry', id: number }>>> };

export type Feed4QueryVariables = Exact<{
  v?: Scalars['String'];
}>;


export type Feed4Query = { __typename?: 'Query', feed?: Maybe<Array<Maybe<{ __typename?: 'Entry', id: number }>>> };

export const FeedDocument = gql\`
    query feed {
  feed {
    id
    commentCount
    repository {
      owner {
        avatar_url
      }
    }
  }
}
    \`;
export const ListenToCommentsDocument = gql\`
    subscription ListenToComments($name: String) {
  commentAdded(repoFullName: $name) {
    id
  }
}
    \`;
export const MutationOperationDocument = gql\`
    mutation MutationOperation($comment: Comment!) {
  submitComment(Comment: $comment) {
    id
  }
}
    \`;
export const Feed2Document = gql\`
    query feed2($v: String!) {
  feed {
    id
  }
}
    \`;
export const Feed3Document = gql\`
    query feed3($v: String) {
  feed {
    id
  }
}
    \`;
export const Feed4Document = gql\`
    query feed4($v: String! = \\"TEST\\") {
  feed {
    id
  }
}
    \`;
export const getSdk = (client: ApolloClient<any>) => ({
      feedQuery(options: Partial<QueryOptions<FeedQueryVariables, FeedQuery>>) {
          return client.query<FeedQuery, FeedQueryVariables>({...options, query: FeedDocument})
      },
listenToCommentsSubscription(options: Partial<SubscriptionOptions<ListenToCommentsSubscriptionVariables, ListenToCommentsSubscription>>) {
          return client.subscribe<ListenToCommentsSubscription, ListenToCommentsSubscriptionVariables>({...options, query: ListenToCommentsDocument})
      },
mutationOperationMutation(options: Partial<MutationOptions<MutationOperationMutation, MutationOperationMutationVariables>>) {
          return client.mutate<MutationOperationMutation, MutationOperationMutationVariables>({...options, mutation: MutationOperationDocument})
      },
feed2Query(options: Partial<QueryOptions<Feed2QueryVariables, Feed2Query>>) {
          return client.query<Feed2Query, Feed2QueryVariables>({...options, query: Feed2Document})
      },
feed3Query(options: Partial<QueryOptions<Feed3QueryVariables, Feed3Query>>) {
          return client.query<Feed3Query, Feed3QueryVariables>({...options, query: Feed3Document})
      },
feed4Query(options: Partial<QueryOptions<Feed4QueryVariables, Feed4Query>>) {
          return client.query<Feed4Query, Feed4QueryVariables>({...options, query: Feed4Document})
      }
    });
    export type SdkType = ReturnType<typeof getSdk>

async function test() {
  const Client = require('@apollo/client').ApolloClient;
  const client = new Client('');
  const sdk = getSdk(client);

  await sdk.feed();
  await sdk.feed3();
  await sdk.feed4();

  const result = await sdk.feed2({ v: \\"1\\" });

  if (result.feed) {
    if (result.feed[0]) {
      const id = result.feed[0].id
    }
  }
}"
`;

exports[`apollo-client sdk Should support useTypeImports 1`] = `
"export type Maybe<T> = T | null;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
import type { ApolloClient, QueryOptions, SubscriptionOptions, MutationOptions } from '@apollo/client';
import gql from 'graphql-tag';
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
};

export type Query = {
  __typename?: 'Query';
  /** A feed of repository submissions */
  feed?: Maybe<Array<Maybe<Entry>>>;
  /** A single entry */
  entry?: Maybe<Entry>;
  /** Return the currently logged in user, or null if nobody is logged in */
  currentUser?: Maybe<User>;
};


export type QueryFeedArgs = {
  type: FeedType;
  offset?: Maybe<Scalars['Int']>;
  limit?: Maybe<Scalars['Int']>;
};


export type QueryEntryArgs = {
  repoFullName: Scalars['String'];
};

/** A list of options for the sort order of the feed */
export enum FeedType {
  /** Sort by a combination of freshness and score, using Reddit's algorithm */
  Hot = 'HOT',
  /** Newest entries first */
  New = 'NEW',
  /** Highest score entries first */
  Top = 'TOP'
}

/** Information about a GitHub repository submitted to GitHunt */
export type Entry = {
  __typename?: 'Entry';
  /** Information about the repository from GitHub */
  repository: Repository;
  /** The GitHub user who submitted this entry */
  postedBy: User;
  /** A timestamp of when the entry was submitted */
  createdAt: Scalars['Float'];
  /** The score of this repository, upvotes - downvotes */
  score: Scalars['Int'];
  /** The hot score of this repository */
  hotScore: Scalars['Float'];
  /** Comments posted about this repository */
  comments: Array<Maybe<Comment>>;
  /** The number of comments posted about this repository */
  commentCount: Scalars['Int'];
  /** The SQL ID of this entry */
  id: Scalars['Int'];
  /** XXX to be changed */
  vote: Vote;
};


/** Information about a GitHub repository submitted to GitHunt */
export type EntryCommentsArgs = {
  limit?: Maybe<Scalars['Int']>;
  offset?: Maybe<Scalars['Int']>;
};

/**
 * A repository object from the GitHub API. This uses the exact field names returned by the
 * GitHub API for simplicity, even though the convention for GraphQL is usually to camel case.
 */
export type Repository = {
  __typename?: 'Repository';
  /** Just the name of the repository, e.g. GitHunt-API */
  name: Scalars['String'];
  /** The full name of the repository with the username, e.g. apollostack/GitHunt-API */
  full_name: Scalars['String'];
  /** The description of the repository */
  description?: Maybe<Scalars['String']>;
  /** The link to the repository on GitHub */
  html_url: Scalars['String'];
  /** The number of people who have starred this repository on GitHub */
  stargazers_count: Scalars['Int'];
  /** The number of open issues on this repository on GitHub */
  open_issues_count?: Maybe<Scalars['Int']>;
  /** The owner of this repository on GitHub, e.g. apollostack */
  owner?: Maybe<User>;
};

/** A user object from the GitHub API. This uses the exact field names returned from the GitHub API. */
export type User = {
  __typename?: 'User';
  /** The name of the user, e.g. apollostack */
  login: Scalars['String'];
  /** The URL to a directly embeddable image for this user's avatar */
  avatar_url: Scalars['String'];
  /** The URL of this user's GitHub page */
  html_url: Scalars['String'];
};

/** A comment about an entry, submitted by a user */
export type Comment = {
  __typename?: 'Comment';
  /** The SQL ID of this entry */
  id: Scalars['Int'];
  /** The GitHub user who posted the comment */
  postedBy: User;
  /** A timestamp of when the comment was posted */
  createdAt: Scalars['Float'];
  /** The text of the comment */
  content: Scalars['String'];
  /** The repository which this comment is about */
  repoName: Scalars['String'];
};

/** XXX to be removed */
export type Vote = {
  __typename?: 'Vote';
  vote_value: Scalars['Int'];
};

export type Mutation = {
  __typename?: 'Mutation';
  /** Submit a new repository, returns the new submission */
  submitRepository?: Maybe<Entry>;
  /** Vote on a repository submission, returns the submission that was voted on */
  vote?: Maybe<Entry>;
  /** Comment on a repository, returns the new comment */
  submitComment?: Maybe<Comment>;
};


export type MutationSubmitRepositoryArgs = {
  repoFullName: Scalars['String'];
};


export type MutationVoteArgs = {
  repoFullName: Scalars['String'];
  type: VoteType;
};


export type MutationSubmitCommentArgs = {
  repoFullName: Scalars['String'];
  commentContent: Scalars['String'];
};

/** The type of vote to record, when submitting a vote */
export enum VoteType {
  Up = 'UP',
  Down = 'DOWN',
  Cancel = 'CANCEL'
}

export type Subscription = {
  __typename?: 'Subscription';
  /** Subscription fires on every comment added */
  commentAdded?: Maybe<Comment>;
};


export type SubscriptionCommentAddedArgs = {
  repoFullName: Scalars['String'];
};
export type FeedQueryVariables = Exact<{ [key: string]: never; }>;


export type FeedQuery = { __typename?: 'Query', feed?: Maybe<Array<Maybe<{ __typename?: 'Entry', id: number, commentCount: number, repository: { __typename?: 'Repository', owner?: Maybe<{ __typename?: 'User', avatar_url: string }> } }>>> };

export type ListenToCommentsSubscriptionVariables = Exact<{
  name?: Maybe<Scalars['String']>;
}>;


export type ListenToCommentsSubscription = { __typename?: 'Subscription', commentAdded?: Maybe<{ __typename?: 'Comment', id: number }> };

export type MutationOperationMutationVariables = Exact<{
  comment: Comment;
}>;


export type MutationOperationMutation = { __typename?: 'Mutation', submitComment?: Maybe<{ __typename?: 'Comment', id: number }> };

export type Feed2QueryVariables = Exact<{
  v: Scalars['String'];
}>;


export type Feed2Query = { __typename?: 'Query', feed?: Maybe<Array<Maybe<{ __typename?: 'Entry', id: number }>>> };

export type Feed3QueryVariables = Exact<{
  v?: Maybe<Scalars['String']>;
}>;


export type Feed3Query = { __typename?: 'Query', feed?: Maybe<Array<Maybe<{ __typename?: 'Entry', id: number }>>> };

export type Feed4QueryVariables = Exact<{
  v?: Scalars['String'];
}>;


export type Feed4Query = { __typename?: 'Query', feed?: Maybe<Array<Maybe<{ __typename?: 'Entry', id: number }>>> };

export const FeedDocument = gql\`
    query feed {
  feed {
    id
    commentCount
    repository {
      owner {
        avatar_url
      }
    }
  }
}
    \`;
export const ListenToCommentsDocument = gql\`
    subscription ListenToComments($name: String) {
  commentAdded(repoFullName: $name) {
    id
  }
}
    \`;
export const MutationOperationDocument = gql\`
    mutation MutationOperation($comment: Comment!) {
  submitComment(Comment: $comment) {
    id
  }
}
    \`;
export const Feed2Document = gql\`
    query feed2($v: String!) {
  feed {
    id
  }
}
    \`;
export const Feed3Document = gql\`
    query feed3($v: String) {
  feed {
    id
  }
}
    \`;
export const Feed4Document = gql\`
    query feed4($v: String! = \\"TEST\\") {
  feed {
    id
  }
}
    \`;
export const getSdk = (client: ApolloClient<any>) => ({
      feedQuery(options: Partial<QueryOptions<FeedQueryVariables, FeedQuery>>) {
          return client.query<FeedQuery, FeedQueryVariables>({...options, query: FeedDocument})
      },
listenToCommentsSubscription(options: Partial<SubscriptionOptions<ListenToCommentsSubscriptionVariables, ListenToCommentsSubscription>>) {
          return client.subscribe<ListenToCommentsSubscription, ListenToCommentsSubscriptionVariables>({...options, query: ListenToCommentsDocument})
      },
mutationOperationMutation(options: Partial<MutationOptions<MutationOperationMutation, MutationOperationMutationVariables>>) {
          return client.mutate<MutationOperationMutation, MutationOperationMutationVariables>({...options, mutation: MutationOperationDocument})
      },
feed2Query(options: Partial<QueryOptions<Feed2QueryVariables, Feed2Query>>) {
          return client.query<Feed2Query, Feed2QueryVariables>({...options, query: Feed2Document})
      },
feed3Query(options: Partial<QueryOptions<Feed3QueryVariables, Feed3Query>>) {
          return client.query<Feed3Query, Feed3QueryVariables>({...options, query: Feed3Document})
      },
feed4Query(options: Partial<QueryOptions<Feed4QueryVariables, Feed4Query>>) {
          return client.query<Feed4Query, Feed4QueryVariables>({...options, query: Feed4Document})
      }
    });
    export type SdkType = ReturnType<typeof getSdk>

async function test() {
  const Client = require('@apollo/client').ApolloClient;
  const client = new Client('');
  const sdk = getSdk(client);

  await sdk.feed();
  await sdk.feed3();
  await sdk.feed4();

  const result = await sdk.feed2({ v: \\"1\\" });

  if (result.feed) {
    if (result.feed[0]) {
      const id = result.feed[0].id
    }
  }
}"
`;
